[quant][core][improvement][feature] Enabled support for quantized fill of nhwc tensors

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #79025

Summary:
Previously, filling a quantized tensor only worked for nchw tensors.
This PR enables support for nhwc tensors. Test cases were added for per
tensor and per channel quantized NHWC tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_fill_per_channel_nhwc
python test/test_quantization.py -k test_qtensor_fill_per_tensor_nhwc
```

Differential Revision: [D36975846](https://our.internmc.facebook.com/intern/diff/D36975846)

dzdang committed:
[quant][core][improvement][feature] Enabled support for quantized fill of nhwc tensors

Summary:
Previously, filling a quantized tensor only worked for nchw tensors.
This PR enables support for nhwc tensors. Test cases were added for per
tensor and per channel quantized NHWC tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_fill_per_channel_nhwc
python test/test_quantization.py -k test_qtensor_fill_per_tensor_nhwc
```

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/Copy.cpp:
```diff
@@ -18,7 +18,8 @@ Tensor& quantized_copy_from_float_(Tensor& self, const Tensor& src) {
-      self.is_contiguous() && src.is_contiguous(),
+      (self.is_contiguous() && src.is_contiguous()) ||
+      (self.is_contiguous(at::MemoryFormat::ChannelsLast) && src.is_contiguous(at::MemoryFormat::ChannelsLast)),
       "Quantized copy only works with contiguous Tensors");
```
do we need to change the error message?

jerryzh168 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -984,6 +1010,37 @@ def test_qtensor_fill_per_channel(self):
+        qtypes = [torch.qint8, torch.quint8, torch.qint32]
+        vals2fill = [-1, 1, 2**32]  # positive, negative, overflow
+        memory_formats = [torch.contiguous_format, torch.channels_last]
+        devices = ["cpu", "cuda"] if TEST_CUDA else ["cpu"]
```
we have a util function here: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_tensor.py#L70, maybe we can move this function to common_quantization.py and import to this file, since it's repeated here as well: https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L168

dzdang replied:
will call this function and have a separate PR that refactors this to common_quantization

jerryzh168 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -984,6 +1010,37 @@ def test_qtensor_fill_per_channel(self):
+            # reference tensor for comparing q_filled
+            q_ref = torch.quantize_per_channel(ones * val2fill, scales=scales,
+                                               zero_points=zero_points, axis=axis, dtype=qtype)
+            self.assertEqual(q_filled.int_repr(), q_ref.int_repr())
+            self.assertEqual(q_filled.dequantize(), q_ref.dequantize())
+            # Make sure the scale and zero_point don't change
+            self.assertEqual(q_filled.q_per_channel_scales(), scales)
+            self.assertEqual(q_filled.q_per_channel_zero_points(), zero_points)
```
have you tried using assertEqual directly?

dzdang replied:
yeah I have, but it doesn't check for eqauivalence of qparams. e.g.,
```
        x = torch.zeros(4)
        scale1 = 1
        scale2 = 2
        zero_point = 0

        qx = torch.quantize_per_tensor(x, scale=scale1, zero_point=zero_point, dtype=torch.qint8)
        qy = torch.quantize_per_tensor(x, scale=scale2, zero_point=zero_point, dtype=torch.qint8)
        self.assertEqual(qx, qy)
        self.assertEqual(qx.dequantize(), qy.dequantize())
```

would pass even though the scales are different. i think we should have assertEqual check qparams as well

dzdang committed:
Update on "[quant][core][improvement][feature] Enabled support for quantized fill of nhwc tensors"


Summary:
Previously, filling a quantized tensor only worked for nchw tensors.
This PR enables support for nhwc tensors. Test cases were added for per
tensor and per channel quantized NHWC tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_fill_per_channel_nhwc
python test/test_quantization.py -k test_qtensor_fill_per_tensor_nhwc
```

Differential Revision: [D36975846](https://our.internmc.facebook.com/intern/diff/D36975846)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][improvement][feature] Enabled support for quantized fill of nhwc tensors"


Summary:
Previously, filling a quantized tensor only worked for nchw tensors.
This PR enables support for nhwc tensors. Test cases were added for per
tensor and per channel quantized NHWC tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_fill_per_channel_nhwc
python test/test_quantization.py -k test_qtensor_fill_per_tensor_nhwc
```

Differential Revision: [D36975846](https://our.internmc.facebook.com/intern/diff/D36975846)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36975846).

dzdang committed:
Update on "[quant][core][improvement][feature] Enabled support for quantized fill of nhwc tensors"


Summary:
Previously, filling a quantized tensor only worked for nchw tensors.
This PR enables support for nhwc tensors. Test cases were added for per
tensor and per channel quantized NHWC tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_fill_per_channel_nhwc
python test/test_quantization.py -k test_qtensor_fill_per_tensor_nhwc
```

Differential Revision: [D36975846](https://our.internmc.facebook.com/intern/diff/D36975846)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36975846).

dzdang committed:
Update on "[quant][core][improvement][feature] Enabled support for quantized fill of nhwc tensors"


Summary:
Previously, filling a quantized tensor only worked for nchw tensors.
This PR enables support for nhwc tensors. Test cases were added for per
tensor and per channel quantized NHWC tensors.

Test Plan:
```
python test/test_quantization.py -k test_qtensor_fill_per_channel_nhwc
python test/test_quantization.py -k test_qtensor_fill_per_tensor_nhwc
```

Differential Revision: [D36975846](https://our.internmc.facebook.com/intern/diff/D36975846)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36975846).
